ngrinder-controller/src/main/java/org/ngrinder/agent/service/AgentService.java

```diff
@@ -0,0 +1,489 @@
+				continue;
+			}
+			// It's my own agent
+			if (fullRegion.endsWith(myAgentSuffix)) {
+				incrementAgentCount(availUserOwnAgent, region, user.getUserId());
+			} else if (!fullRegion.contains("owned_")) {
+				incrementAgentCount(availShareAgents, region, user.getUserId());
+			}
+		}
+
+		int maxAgentSizePerConsole = agentManager.getMaxAgentSizePerConsole();
+
+		for (String region : regions) {
+			MutableInt mutableInt = availShareAgents.get(region);
+			int shareAgentCount = mutableInt.intValue();
+			mutableInt.setValue(Math.min(shareAgentCount, maxAgentSizePerConsole));
+			mutableInt.add(availUserOwnAgent.get(region));
+		}
+		return availShareAgents;
+	}
+
+	private void incrementAgentCount(Map<String, MutableInt> agentMap, String region, String userId) {
+		if (!agentMap.containsKey(region)) {
+			LOGGER.warn("Region: {} not exist in cluster nor owned by user: {}.", region, userId);
+		} else {
+			agentMap.get(region).increment();
+		}
+	}
+
+	protected Set<String> getRegions() {
+		return regionService.getAll().keySet();
+	}
+
+	@Override
+	public AgentControllerIdentityImplementation getAgentIdentityByIpAndName(String ip, String name) {
+		AgentInfo agentInfo = agentInfoStore.getAgentInfo(createAgentKey(ip, name));
+		if (agentInfo != null) {
+			return cast(agentInfo.getAgentIdentity());
+		}
+		return null;
+	}
+
+	@Override
+	public AgentInfo getOne(String ip, String name) {
+		return agentInfoStore.getAgentInfo(createAgentKey(ip, name));
+	}
+
+	/**
+	 * Approve/disapprove the agent on given id.
+	 *
+	 * @param ip      ip
+	 * @param name	  host name
+	 * @param approve true/false
+	 */
+	@Transactional
+	public void approve(String ip, String name, boolean approve) {
+		AgentInfo agentInfoInDB = agentManagerRepository.findByIpAndName(ip, name);
+
+		if (agentInfoInDB != null) {
+			agentInfoInDB.setApproved(approve);
+		} else {
+			agentInfoInDB = new AgentInfo();
+			agentInfoInDB.setIp(ip);
+			agentInfoInDB.setName(name);
+			agentInfoInDB.setApproved(approve);
+			agentManagerRepository.save(agentInfoInDB);
+		}
+		agentManagerRepository.flush();
+
+		AgentInfo agentInfo = agentInfoStore.getAgentInfo(createAgentKey(ip, name));
+		agentInfo.setApproved(approve);
+		agentInfoStore.updateAgentInfo(agentInfo.getAgentKey(), agentInfo);
+	}
+
+	/**
+	 * Get all free approved agents
+	 *
+	 * @return AgentInfo set
+	 */
+	public Set<AgentInfo> getAllFreeApprovedAgents() {
+		return getAllReady()
+			.stream()
+			.filter(AgentInfo::isApproved)
+			.collect(toSet());
+	}
+
+	/**
+	 * Get all approved agents for given user which are not used now.
+	 *
+	 * @param user user
+	 * @return AgentInfo set
+	 */
+	public Set<AgentInfo> getAllAttachedFreeApprovedAgentsForUser(User user) {
+		if (user == null) {
+			return emptySet();
+		}
+
+		String region = config.getRegion();
+		String userId = user.getUserId();
+		return getAllFreeApprovedAgents()
+			.stream()
+			.filter(AgentInfo::isApproved)
```

Comment:
why do we need isApproved here? getAllFreeApprovedAgents() returns approved agents?

Reply:
Yes, ㅅhis seems unnecessary.